Packages: Add new @wordpress/data-controls package.

nerrad opened the pull request:
## Description
This is a new package triggered in part by [this](https://github.com/WordPress/gutenberg/pull/14634#issuecomment-476627137) and [this comment](https://github.com/WordPress/gutenberg/pull/14634#issuecomment-476628388).

The `@wordpress/data-controls` package exposes a set of common controls consumers can use when registering custom stores.  In the initial iteration the following controls are included:

* `select`: For selecting from another data store in the registry.
* `dispatch`: For dispatching actions to another store in the registry.
* `apiFetch`: For doing a request via the `@wordpress/api-fetch` module.

## Notes:

- The `select` control has been made resolver aware.  Essentially this means if the named selector has a sibling resolver registered in the given store, then a promise is returned that resolves to the value for the selector after any necessary resolution is completed.  If the selector does not have a resolver sibling then it is called directly.  This allows for client code to not have to worry about resolution when using this control (**Note:** This requires a change in `@wordpress/data`: see #15436).
- I've implemented this new package with the `@wordpress/editor`.  Primarily I did this so that things could be tested but I'm on the fence with whether this should be done in its own pull.  So I'm looking from feedback from reviewers on this. It turns out the the changes aren't that much (other than eliminating code 🎉) so I think it'd be good to merge with an initial implementation of the package.

## Todos:

* [x] Write package
* [x] Update main package.json for the mono-repo
* [x] Do pull for `@wordpress/data` exposing whether a selector has a resolver (see #15436)
* [x] Write README.md
* [x] Write CHANGELOG.md
* [x] Update docs builder and make sure auto-generated docs look okay.
* [x] Add tests
* [x] Implement this somewhere (`@wordpress/editor`) to ensure api is usable and works as expected.
* [x] If the above ^^ stays in this branch and is released with this pull, then the appropriate `CHANGELOG.md` and tests will need updated.

## How has this been tested?

* [x] New unit tests added
* [x] Implemented with `@wordpress/editor` and ensure the editor continues to work via user testing.
* [x] e2e tests should still pass with this work.

## Types of changes

- This is a non-breaking change introducing a new package.
- As an example of implementation, this has been implemented with the `core/editor` store.  Thus the editor is impacted but impacted areas are covered by e2e and unit tests.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

nerrad committed:
Add new @wordpress/data-controls package

nerrad committed:
include new package in main package.json

nerrad committed:
Implement new package with @wordpress/editor store

- this might be extracted to its own pull but added initially for testing.

nerrad committed:
docs tool generating this for some reason

nerrad committed:
include new package with doc generation tool

nerrad committed:
updated manifest from doc generation

nerrad committed:
updated package-lock.json

nerrad commented on packages/data-controls/src/index.js:
```diff
@@ -0,0 +1,171 @@
+			if ( registry.select( storeKey )[ selectorName ].hasResolver ) {
+				return resolveSelect( registry, { storeKey, selectorName, args } );
+			}
+			return registry.select( storeKey )[ selectorName ]( ...args );
```
I can probably just do a ternary for this.

nerrad committed:
not sure why docs build tool is doing this.

nerrad committed:
use ternary

nerrad committed:
data-controls added to manifest by docs build tool

nerrad committed:
update README for editor package

nerrad committed:
updates to editor store actions tests

- to use new data-controls package

nerrad committed:
add missing `@example` tag and regenerate docs

nerrad committed:
add tests for package

nerrad committed:
fix typo in changelog for package name

nerrad committed:
fix typo for package name in readme

nerrad committed:
more typos

gziolo commented on bin/update-readmes.js:
```diff
@@ -17,6 +17,7 @@ const packages = [
 		'Autogenerated selectors': 'src/selectors.js',
 	} ],
 	'data',
+	'data-controls',
```
@nosolosw, have you had a chance to revisit this list so we could use blacklist instead to make it easier to add new packages?

nosolosw replied:
It has grown on me that an allow list works better in this case for two reasons:

- I know we've talked about detecting the entry point from package.json. Note, however, how some packages will need more than this (example: `core-data`. This will require an allow list anyway.
- Using a deny list will do unnecessary extra work. Currently, docgen will report an error if it doesn't find a token in a README file, but this can be improved to gracefully continue. However, we'll need parsing the package entry point files anyway to know whether there is a token or not - wasting those cycles.

Adding/Removing a package is not a common operation, so I think is fine to require authors to do this explicitely if they want API doc generation.

gziolo commented on packages/data-controls/package.json:
```diff
@@ -0,0 +1,31 @@
+{
+	"name": "@wordpress/data-controls",
+	"version": "1.0.0",
```
```suggestion
	"version": "1.0.0-beta.1",
```
See: https://github.com/WordPress/gutenberg/tree/19d1cbfd43daeaf53fc9b3987e2c21eff54aab4b/packages#creating-a-new-package.

Lerna bumps version on every release so we need to start with a version lower than `1.0.0`.

gziolo commented on packages/data-controls/README.md:
```diff
@@ -0,0 +1,134 @@
+
+The following controls are available on the object returned by the module:
+
+<!-- START TOKEN(Autogenerated API docs) -->
```
```suggestion
## API

<!-- START TOKEN(Autogenerated API docs) -->
```

gziolo reviewed (commented):
I guess this should be reviewed by @youknowriad or @aduth. I can try myself but I didn't follow recent changes in this area.

I left my comments related to the package itself. Everything looks great, there are some tiny things to updated which I highlighted.

nerrad committed:
Use  lower version for initial release.

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

nerrad committed:
Add heading for API.

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

aduth commented:
First thought: Is `apiFetch` too specific to WordPress use case for broad applicability in such a package? I could imagine `fetch` being useful, and I forget the exact reasons `@wordpress/api-fetch` is not marketed as a generic fetch utility.

aduth commented:
Thought: Should these (at least `select`, `dispatch`) be offered by the `data` package itself? Possibly also by default? The more I think on it, the more I think "No" (prefer opt-in), but mentioning as it occurs to me. Similarly there's the thought of respective packages offering controls (`api-fetch`), but I also think this bloats the scope of the modules to become aware of the data package and its specific mechanics.

aduth commented:
>This allows for client code to not have to worry about resolution when using this control 

I like this enhancement 👍

aduth reviewed (approved):
I'd be more okay to start with just `select` and `dispatch`, but I could be convinced of `apiFetch` if the arguments are made. The code looks good.

aduth reviewed (changes requested):
We need to update this, since it has a default export. Otherwise the window global doesn't work as expected (`wp.dataControls.default`):

https://github.com/WordPress/gutenberg/blob/8960509d97f72bf4bee68e1513a6c3fd3b0d446e/webpack.config.js#L71-L78

Although, this is a unique case of needing both default and named exports. How do you propose this might be handled from the global? Is referencing the controls object by `.default` acceptable? It doesn't seem quite so obvious to me.

nerrad commented:
> I'd be more okay to start with just select and dispatch, but I could be convinced of apiFetch if the arguments are made.

Arguments for:

- arbitrary requests via api is one of the more common use-cases for a control.  Having a default utility in place for this is helpful for the end user.
- The package is only useful for consumers of the `@wordpress/data` package and registries implementing the controls plugin (available by default on the exposed defaultRegistry).  Thus it makes sense that a fetch control would use the api fetch utility from the `@wordpress` namespace.
- There is immediate benefit within the GB project using this library as nearly always these three controls will be sufficient.
- exposing these controls does not preclude implementors overriding or supplementing with their own controls.  It just helps expose a set of sensible defaults that in most cases will be sufficient.

> Although, this is a unique case of needing both default and named exports. How do you propose this might be handled from the global? Is referencing the controls object by .default acceptable? It doesn't seem quite so obvious to me.

Interesting, this was a specific about the build process I wasn't aware of.  I'm also a bit surprised that the implementation here of the new package (in @wordpress/editor) didn't break because of that.

Typical usage will be importing the default for registering with the store (as done [here](https://github.com/WordPress/gutenberg/blob/8ce6b53254d9d0ac359aec0a7440e68097591f8b/packages/editor/src/store/index.js#L5)) Are you thinking it'd be better to export it as a named export so that it's a bit more evident it's purpose on the global (i.e. `wp.dataControls.controls`)?

nerrad committed:
Update CHANGELOG.md version reference.

Follows repo conventions about using Master as the version heading for unreleased changes.

nerrad commented:
Oh I see, the `LibraryExportDefaultPlugin` exports defaults onto the named global (eg. `wp.apiFetch` as opposed to `wp.apiFetch.default`)?  I don't use ES5 hardly at all so I don't notice these things.  So what happens here.  So to clarify, this means:

**With configuration in webpack plugin**

- `wp.dataControls` for default
- `wp.dataControls.select`, `wp.dataControls.dispatch`... etc for named exports?

**Without configuration in webpack plugin**

- `wp.dataControls.default` for default
- `wp.dataControls.select`, `wp.dataControls.dispatch` ...etc for named exports?

nerrad committed:
add data-controls to webpack plugin config

nerrad committed:
Revert "add data-controls to webpack plugin config"

This reverts commit 36a0b98acc6d81e2921c09db5e20c1ad68e28e47.

nerrad commented:
So I tried adding this to the plugin and it looks like because the default export is an object that the named exports are not accessible on that object.  So the choices currently are:

1. Leave the default export as `wp.dataControls.default`.
2. Export the controls object as a named export (maybe `wp.dataControls.controls`?)

Personally I'm fine with leaving it as `wp.dataControls.default` as in most cases I imagine this will be used in ES6 code anyways.   However, I'm open to exporting as a named export.

youknowriad commented:
I wonder if 

 - having all the controls in a single export 
 - and at the same time, having the actions in separate exports 

is a good way to go about this?

 - Shouldn't we have two exported variables: `controls` and `actions`?
 - Alternatively, we could consider whether we export things like:

```
export const select = { 
   control: selectControl,
   action: selectAction
}
```

I don't have the right answer personally

nerrad commented:
> Shouldn't we have two exported variables: controls and actions?

I could see doing something like that because it retains the ability to register the store (a la)

```
// es6
import { controls } from '@wordpress/data-controls'
// es5
const { controls } = wp.dataControls;
```

And the actions are still exposed in a somewhat friendly format (one more step for to get the specific actions though):

```
// es6
import { actions } from '@wordpress/data-controls';
const { select, dispatch } = actions;

//es5
const { actions } = wp.dataControls
const { select, dispatch, apiFetch } = actions
```

However I don't think it'd be a good idea to have the control and action on each object key value as it makes it less intuitive for setting up what's needed for the different contexts.

Personally, I'm still okay with either `wp.dataControls.default, wp.dataControls,select, ...etc` or `wp.dataControls.controls, wp.dataControls.select, ...etc`.

youknowriad commented:
> wp.dataControls.controls, wp.dataControls.select, ...etc.

I'd go with this one personally in that case

nerrad committed:
use named export for controls object

nerrad committed:
update docs

nerrad commented:
9003bc5 exports the controls configuration object on a named export.  So now the global has:

- `wp.dataControls.controls` for the controls object.
- `wp.dataControls.select`
- `wp.dataControls.dispatch`
- `wp.dataControls.apiFetch`

youknowriad commented:
Looks like the tests are broken, otherwise, this is looking good.

nerrad committed:
woop fix import on tests

aduth commented:
I agree `controls` as a named export seems sensible (arguably even preferable, since I don't know that it's as intuitive as the default export).

jorgefilipecosta commented:
Merging this PR as it seems ready and was approved, and I have a PR depending on it that I would like to merge. If there are is any other improvement I think it can be shipped in a follow-up PR.

jorgefilipecosta commented:
> @jorgefilipecosta since you've implemented this in another branch, do you have a preference or some ideas here?

Sorry, I missed this question, personally like the version that was proposed:
wp.dataControls.controls for the controls object.
wp.dataControls.select
wp.dataControls.dispatch
wp.dataControls.apiFetch
